### PR TITLE
Vulkan: Implement framebuffer "depal" (palette lookup) and shader blending

### PR DESCRIPTION
**Pull request by hrydgard**
Fixes outlines in DBZ: Tenkachi, shadows in FF:Type0, and many other various effects.

Does not do any special Vulkan magic for shader blending (input attachments) though that might be worth trying at a later time.

### CODE REVIEW

File: Windows/GPU/WindowsVulkanContext.cpp [hunk 1]
```diff
@@ -128,6 +128,11 @@ static VkBool32 VKAPI_CALL Vulkan_Dbg(VkDebugReportFlagsEXT msgFlags, VkDebugRep
 		return false;
 	if (msgCode == 64)  // Another useless perf warning that will be seen less and less as we optimize -  vkCmdClearAttachments() issued on command buffer object 0x00000195296C6D40 prior to any Draw Cmds. It is recommended you use RenderPass LOAD_OP_CLEAR on Attachments prior to any Draw.
 		return false;
+	
+	// TEMPORARy
+	if (msgCode == 6)
```
Comment (hrydgard) [inline Windows/GPU/WindowsVulkanContext.cpp]: Oops, gonna remove this.

File: GPU/Vulkan/TextureCacheVulkan.cpp [hunk 2]
```diff
@@ -400,27 +399,43 @@ void TextureCacheVulkan::ApplyTextureFramebuffer(TexCacheEntry *entry, VirtualFr
...
+		draw_->BindFramebufferAsTexture(framebuffer->fbo, 0, Draw::FB_COLOR_BIT, 0);
+		VkImageView fbo = (VkImageView)draw_->GetNativeObject(Draw::NativeObject::BOUND_TEXTURE0_IMAGEVIEW);
 
-		//depalFBO->EndPass(cmd);
-		//depalFBO->TransitionToTexture(cmd);
-		//imageView = depalFBO->GetColorImageView();
+		VkDescriptorSet descSet = vulkan2D_->GetDescriptorSet(fbo, samplerNearest_, clutTexture->GetImageView(), samplerNearest_);
+		VulkanRenderManager *renderManager = (VulkanRenderManager *)draw_->GetNativeObject(Draw::NativeObject::RENDER_MANAGER);
+		renderManager->BindPipeline(depalShader->pipeline);
+		renderManager->SetScissor(VkRect2D{ {0, 0}, { framebuffer->renderWidth, framebuffer->renderHeight} });
```
Comment (unknownbrackets) [inline GPU/Vulkan/TextureCacheVulkan.cpp]: Do we dirty this state afterward? I may have missed it.

-[Unknown]

File: GPU/Vulkan/DepalettizeShaderVulkan.h [hunk 3]
```diff
@@ -20,42 +20,58 @@
...
-	GLuint fragShader;
-	GLint a_position;
-	GLint a_texcoord0;
-	*/
+	~DepalShaderVulkan() {
+		delete[] code;
+	}
+	// A Vulkan2D pipeline. Set texture to slot 0 and palette texture to slot 1.
+	VkPipeline pipeline = VK_NULL_HANDLE;
+	const char *code = nullptr;;
```
Comment (unknownbrackets) [inline GPU/Vulkan/DepalettizeShaderVulkan.h]: nit: `;;`.

-[Unknown]

**Comment by unknownbrackets** [comment]
Brave Story - my goto test for shader blending - doesn't work yet.

This is the one that sequentially draws over itself for some effects (like a less evil Sonic Rivals 2.)  When I stepped through the entire frame using the GE debugger - it did appear correctly.  This is exactly what happened in GLES, and it was because render-to-self wasn't working right...

3rd Birthday looks right though, cool.  Still testing some others...

-[Unknown]

**Comment by hrydgard** [comment]
Yes, I've noticed some problems with shader blending too, it seems we end up with the wrong primary texture bound, looking at it in RenderDoc and comparing with D3D11. I'll debug it soon.

File: GPU/Vulkan/TextureCacheVulkan.cpp (hunk 2 above)
Comment (hrydgard) [reply GPU/Vulkan/TextureCacheVulkan.cpp]: Nope. Fixing, thanks.

**Comment by unknownbrackets** [comment]
Brave Story isn't an issue with shader blending or etc.  I forgot that when logic ops are supported, it doesn't use it.

The good news is if I force disable GPU_SUPPORTS_LOGIC_OP, it works.  So everything is good with the shader stuff - it's logic ops which are broken.

-[Unknown]

**Pull request merged**